### DISCUSSION TITLE
Label targets

### DISCUSSION THREAD

**Discussion by JouriM2** [AvaloniaUI/Avalonia#15939] [Q&A] [answered]
Here is sample repo: https://github.com/JouriM2/AvDropDownBug
It contains simple window with 3 edits and labels to access them.

![изображение](https://github.com/AvaloniaUI/Avalonia/assets/114450272/8953f961-1075-4bdc-907c-f28a8d6ce41d)

implemented as:
```XML
	<StackPanel Orientation="Vertical" Margin="20">
		<Label Content="_1 One" Target="edt1"/>
		<TextBox x:Name="edt1"/>

		<Label Content="_2 Two" Target="edt2"/>
		<TextBox x:Name="edt2"/>

		<Label Content="_A Three" Target="edt3"/>
		<TextBox x:Name="edt3"/>

	</StackPanel>
```

Problems I have:

1. Number hotkeys not work. Pressing Alt+1 or Alt+2 do nothing. Alt+A works as expected and move focus to edit.
  If to press mouse on `Label` itself all three are works.
  How to use number hotkeys?

2. I need hotkey allways be underlined, not only after Alt pressed. How to get it?

3. After I press Alt on keyboard all hotkeys stay underlined until I press any mouse button inside window (any button, at any point inside window).
  At same time any keyboard actions do not hemove highlighting.
  It seems like a bug and hotkeys need to be hidden after Alt key released.

**Comment by stevemonaco** [comment, marked as answer]
For point 1, this is almost certainly a bug in this area and should be reported: https://github.com/AvaloniaUI/Avalonia/blob/f140033e420ee39b16dcc0621d5580f94d5bcb2c/src/Avalonia.Base/Input/AccessKeyHandler.cs#L175-L183

`e.Key.ToString()` returns `"D1"` for the key (because of the enum) whereas it's registered as "1" in the map.

For point 2, you can use a selector and probably get away with it:
```xml
<Style Selector="Label AccessText">
    <Setter Property="ShowAccessKey" Value="True" />
</Style>
```

For point 3, it sort of conflicts with the "always showing" part of 2. Anyways, this works as expected if you have a `Menu`, which is the primary use of `AccessText`. I'm not sure what the intended design is for out-of-menu access keys. You should be able to press the alt key, release it, and continue pressing access keys (with or without alt held down) to navigate. The non-menu access keys return to normal when the menu is closed.

It can be fixed in this area https://github.com/AvaloniaUI/Avalonia/blob/f140033e420ee39b16dcc0621d5580f94d5bcb2c/src/Avalonia.Base/Input/AccessKeyHandler.cs#L211-L230 by placing the following before the `break;`. Not sure if there are any side-effects, but I think this warrants a bug issue.

```cs
if (MainMenu is null)
{
    _owner!.ShowAccessKeys = _showingAccessKeys = false;
}
```

**Reply by JouriM2** [reply]
1. reported (if I do it right)

2. Thanks. It works.
Its very complex to write selector for each label, but I have my own theme so I change global `Label` template.
May be it will be usefull to someone.
To make non-class selectors in control template we neet to write something like this:
```XML
	<ControlTheme x:Key="{x:Type Label}" TargetType="Label">
...
		<Setter Property="Template">
			<ControlTemplate>
				<ContentPresenter Name="PART_ContentPresenter"
...
					<ContentPresenter.Styles>
						<Style Selector="AccessText">
							<Setter Property="ShowAccessKey" Value="True"/>
						</Style>
					</ContentPresenter.Styles>
				</ContentPresenter>
...
```
3. I mean I personally dont need to hide access keys but I think **IF** its used it must be show only until Alt key is hold.
But may be I get the idea wrong, never use this option (to hide access keys) in my life.
Plus.
`Menu` not only and not even main "user" of hot-keys.
I trying to use Avalonia as a replacement for my existing GUI toolkit (which is Windows only) to create **desktop** applications.
For me (and users of my applications) desktop is not about "looking great, just like on my phone", but about "how fast I can work in complex GUI". It includes mostly with keyboard. In this situation hot-key is a big deal to allow fast navigftion from one GUI part to another without using mouse.
